KAFKA-8286; Leader Election Admin RPC (KIP-460)

jsancio opened the pull request:
Implements KIP-460: https://cwiki.apache.org/confluence/display/KAFKA/KIP-460%3A+Admin+Leader+Election+RPC

Pending:
1. Implement additional tests
1. Improve documentation

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

jsancio committed:
Change the elect leader RPC

jsancio committed:
Election type support in the controller

jsancio committed:
Implement kafka-elect-leaders command

jsancio committed:
Add PartitionLeaderNotAvailable exception

Also remove any remaining references to "preferred".

jsancio committed:
Add tests for partition state machine

jsancio committed:
Merge the preferred election command test suites

jsancio committed:
Add framework for testing LeaderElectionCommand

jsancio commented:
cc @hachikuji @mumrah @junrao

mumrah commented on clients/src/main/java/org/apache/kafka/common/ElectionType.java:
```diff
@@ -0,0 +1,36 @@
+ */
+@InterfaceStability.Evolving
+public enum ElectionType {
+    PREFERRED((byte) 0), UNCLEAN((byte) 1);
```
What other election types do we foresee?

jsancio replied:
The controller supports 5 types of leader elections so we may want to expose those through this RPC in the future. e.g. Elect a "better" (preferred not required but not unclean) leader.

mumrah commented on clients/src/main/java/org/apache/kafka/common/errors/PartitionLeaderNotAvailableException.java:
```diff
@@ -0,0 +1,28 @@
+ */
+package org.apache.kafka.common.errors;
+
+public class PartitionLeaderNotAvailableException extends InvalidMetadataException {
```
Do we need this instead of `LeaderNotAvailableException`? Is there a semantic difference between the two?

jsancio replied:
I opted not to use `LeaderNotAvailableException` as it is mainly used in the fetcher code and it has a different meaning. I renamed the exception to `EligibleLeadersNotAvailableException`

mumrah commented on clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java:
```diff
@@ -190,8 +190,8 @@ public Struct parseResponse(short version, ByteBuffer buffer) {
     DELETE_GROUPS(42, "DeleteGroups", DeleteGroupsRequest.schemaVersions(), DeleteGroupsResponse.schemaVersions()),
-    ELECT_PREFERRED_LEADERS(43, "ElectPreferredLeaders", ElectPreferredLeadersRequestData.SCHEMAS,
-            ElectPreferredLeadersResponseData.SCHEMAS),
+    ELECT_LEADERS(43, "ElectLeaders", ElectLeadersRequestData.SCHEMAS,
```
Is this safe? Can we rename APIs without side effects?

hachikuji replied:
The API name is not technically a public interface. We did a similar renaming from GroupCoordinator request to FindCoordinator. That said, this probably will break some downstream systems that were using internals anyway. I suspect our leeway to do stuff like this gets smaller with the schema definitions represented in JSON. Still, I think we should do it.

jsancio replied:
Yes. This in internal to the broker. The protocol only uses the code. E.g. 43. Can you confirm @hachikuji  or @cmccabe ?

mumrah commented on clients/src/main/java/org/apache/kafka/common/requests/ElectLeadersRequest.java:
```diff
@@ -0,0 +1,132 @@
+import org.apache.kafka.common.protocol.types.Struct;
+import org.apache.kafka.common.utils.CollectionUtils;
+
+public class ElectLeadersRequest extends AbstractRequest {
```
I think the new way to add APIs is using the JSON schema notation and the schema generator.

jsancio replied:
Correct, the type generated is `ElectLeadersRequestData`. This type is a wrapper around that type to make it easier to create, to deal with different versions and to interact with the transport layer.

mumrah commented on core/src/main/scala/kafka/admin/PreferredReplicaLeaderElectionCommand.scala:
```diff
@@ -42,6 +42,7 @@ object PreferredReplicaLeaderElectionCommand extends Logging {
     run(args, timeout)
   }
   def run(args: Array[String], timeout: Int = 30000): Unit = {
+    println("This tool is deprecate. Please use kafka-leader-election tool. Tracking issue at: TBD")
```
"deprecated"

hachikuji replied:
typo: deprecated?

mumrah commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -29,7 +29,9 @@ import kafka.utils._
 import kafka.zk._
 import kafka.zookeeper.{StateChangeHandler, ZNodeChangeHandler, ZNodeChildChangeHandler}
-import org.apache.kafka.common.{KafkaException, TopicPartition}
+import org.apache.kafka.common.ElectionType
```
These three imports can be combined with `{ }` (I think that's the preferred style with scala)

jsancio replied:
I try to avoid that style as it makes it more difficult to deal with merge conflicts when adding or removing unrelated imports.

mumrah commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -635,38 +641,50 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
-  object AutoTriggered extends ElectionType
-  object ZkTriggered extends ElectionType
-  object AdminClientTriggered extends ElectionType
+  sealed trait ElectionTrigger
```
👍 for better name

mumrah commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1551,12 +1569,20 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
+  case class ReplicaLeaderElection(
+    partitionsFromAdminClientOpt: Option[Set[TopicPartition]],
+    electionType: ElectionType,
+    electionTrigger: ElectionTrigger = ZkTriggered,
```
Should we keep this default here, or force callers to be explicit?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/Errors.java:
```diff
@@ -305,7 +306,9 @@
     FENCED_INSTANCE_ID(82, "The coordinator reports a more recent member.id associated with the consumer's group.instance.id.",
-            FencedInstanceIdException::new);
+            FencedInstanceIdException::new),
+    PARTITION_LEADER_NOT_AVAILABLE(83, "A topic partition leader was not available",
```
Could we not use `LEADER_NOT_AVAILABLE`?

jsancio replied:
Yeah. I though about using that error code. The felt that the string message reported by the protocol is miss leading. Looking at the code, this error is mainly used by the broker to report errors during a fetch. Here we are trying to report that the controller couldn't elect a leader because all of the replicas were unavailable.

I am okay reusing the error and/or making the string message a bit more generic. Let me know what you think.

hachikuji replied:
Yeah, the use case is a bit different. Perhaps we just need a different name. Maybe `NO_ELIGIBLE_LEADERS_AVAILABLE` or something like that?

jsancio committed:
Close AdminClient after elect leader command

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -848,7 +849,9 @@ public DeleteConsumerGroupsResult deleteConsumerGroups(Collection<String> groupI
      *
      * @param partitions      The partitions for which the preferred leader should be elected.
      * @return                The ElectPreferredLeadersResult.
+     * @deprecated            Since TBD. Use {@link #electLeaders}.
```
I think we can list this as 2.3. It's probably worth mentioning this in the upgrade notes as well.

jsancio replied:
Got it. Can you link me to the upgrade notes?

hachikuji replied:
It's at `docs/upgrade.html`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ElectLeadersOptions.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.Collection;
+
+/**
+ * Options for {@link AdminClient#electLeaders(ElectionType, Collection, ElectLeadersOptions)}.
```
nit: seems this link is broken

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ElectLeadersOptions.java:
```diff
@@ -0,0 +1,31 @@
+ * The API of this class is evolving, see {@link AdminClient} for details.
+ */
+@InterfaceStability.Evolving
+final public class ElectLeadersOptions extends AbstractOptions<ElectLeadersOptions> {
```
Not sure it matters too much, but the other `Options` classes are not final. Was there a reason to make this one final?

jsancio replied:
It is a habit. JVM can make dispatch optimization if it has a value of type `ElectLeadersOptions` since it know that it is not possible for the virtual method table to be overridden by a subtype.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ElectLeadersResult.java:
```diff
@@ -0,0 +1,137 @@
+import org.apache.kafka.common.requests.ApiError;
+
+/**
+ * The result of {@link AdminClient#electLeaders(ElectionType, Collection, ElectLeadersOptions)}
```
nit: link broken

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ElectLeadersResult.java:
```diff
@@ -0,0 +1,137 @@
+                    result.completeExceptionally(new UnknownTopicOrPartitionException(
+                            "Leader election for partition \"" + partition +
+                                    "\" was not attempted"));
+                } else if (partitions == null && topicPartitions.isEmpty()) {
```
This is from the original, but Intellij complains here. If `topicPartitions.containsKey(partition)` is true, then `topicPartitions.isEmpty()` must be false. Should this check go earlier? We might be missing a test case.

By the way, this seems like a very obscure way to infer a cluster authorization failure. Isn't it possible that the cluster simply has no topics? Perhaps we should add a top-level error code to the response since we are bumping the protocol anyway?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ElectLeadersResult.java:
```diff
@@ -0,0 +1,137 @@
+                if (throwable != null) {
+                    result.completeExceptionally(throwable);
+                } else if (!topicPartitions.containsKey(partition)) {
+                    result.completeExceptionally(new UnknownTopicOrPartitionException(
```
If the partition was not in the result, that seems like an unexpected error. I wonder if it is more appropriate to use a vanilla `KafkaException`.

jsancio replied:
To me, I think that the question is if we expect the client to handle this exception. You are right, this is probably a program error and we should instead just throw the generic `KafkaException`. I will change this.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2389,32 +2401,31 @@ class KafkaApis(val requestChannel: RequestChannel,
+        new ElectLeadersResponse(data)})
     }
     if (!authorize(request.session, Alter, Resource.ClusterResource)) {
       val error = new ApiError(Errors.CLUSTER_AUTHORIZATION_FAILED, null);
```
nit: unneeded semicolon

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ElectPreferredLeadersResult.java:
```diff
@@ -34,9 +34,12 @@
 @InterfaceStability.Evolving
-public class ElectPreferredLeadersResult {
+@Deprecated
+final public class ElectPreferredLeadersResult {
```
I wonder if there's any way we can extend `ElectLeadersResult` to avoid the duplication. We'll have to fix the bugs twice. 

Also, I think the addition of `final` breaks compatibility.

jsancio replied:
I'll remove the `final`.

Regarding code duplication. We have 3 options:
1. `ElectLeadersResult extends ElectPreferredLeadersResult`. This means that a supported type is a deprecated type. Eventually, we want to remove `ElectPreferredLeadersResult` which means the type hierarchy for `ElectLeadersResult` will change in a backwards incompatible way.
2. `ElectPreferredLeadersResult extends ElectLeadersResult`. This is a change in the type hierarchy for `ElectPreferredLeadersResult` but maybe this is okay. This also means that `ElectLeadersResult` is not final which I am trying to avoid.
3. Share some code through composition. I am not 100% sure how this would work but let me investigate. Maybe `ElectPreferredLeadersResult` can be a wrapper for `ElectLeadersResult` that delegates to it.

jsancio replied:
`electPreferredLeaders` now uses `electLeaders`. I implemented `ElectPreferredLeadersResult` using option 3.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ElectLeadersResult.java:
```diff
@@ -0,0 +1,137 @@
+
+    /**
+     * <p>Get a future for the topic partitions for which a leader election
+     * was attempted. A partition will be present in this result if
```
Is this true? Don't we call `completeExceptionally` on partition failures below?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ElectLeadersResult.java:
```diff
@@ -0,0 +1,137 @@
+     */
+    public KafkaFuture<Set<TopicPartition>> partitions() {
+        if (partitions != null) {
+            return KafkaFutureImpl.completedFuture(this.partitions);
```
I know this isn't from this patch, but I'm having a hard time understanding this API. I wonder if we can make the behavior consistent whether the operation is attempted on all partitions or not.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ElectLeadersResult.java:
```diff
@@ -0,0 +1,137 @@
+        final KafkaFutureImpl<Void> result = new KafkaFutureImpl<>();
+        electionFuture.thenApply(new KafkaFuture.Function<Map<TopicPartition, ApiError>, Void>() {
+            @Override
+            public Void apply(Map<TopicPartition, ApiError> topicPartitions) {
```
We don't have the `topicPartitions.isEmpty` check here, so we wouldn't see the cluster authorization error that we would see through the other futures.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2943,34 +2944,66 @@ void handleFailure(Throwable throwable) {
         final long now = time.milliseconds();
-        runnable.call(new Call("electPreferredLeaders", calcDeadlineMs(now, options.timeoutMs()),
+
+        runnable.call(new Call("electLeaders", calcDeadlineMs(now, options.timeoutMs()),
```
As far as I can tell, this call is the same as the one in `electLeaders` except the election type is statically defined. It's not a lot of code, but is it worth factoring it out?

jsancio replied:
Yes. I'll revisit these set of changes and update the PR.

hachikuji reviewed (commented):
Thanks for the patch. Left a few initial comments.

hachikuji commented on core/src/main/scala/kafka/admin/LeaderElectionCommand.scala:
```diff
@@ -0,0 +1,235 @@
+    val commandOptions = new LeaderElectionCommandOptions(args)
+    CommandLineUtils.printHelpAndExitIfNeeded(
+      commandOptions,
+      "This tool attempts to elect a new leader from a set of topic partitions. The type of elections supported are preferred replicas and unclean replicas."
```
Perhaps "*for* a set of topic partitions"?

hachikuji commented on core/src/main/scala/kafka/admin/LeaderElectionCommand.scala:
```diff
@@ -0,0 +1,235 @@
+  val electionType = parser
+    .accepts(
+      "election-type",
+      "Type of election to attempt. Possible values are \"preferred\" for preferred leader election or \"unclean\" for unclean leader election. REQUIRED.")
```
It might be worth mentioning somewhere that unclean leader election will only occur if the partition is offline (i.e. has no leader).

hachikuji commented on core/src/main/scala/kafka/admin/LeaderElectionCommand.scala:
```diff
@@ -0,0 +1,235 @@
+      "topic",
+      "Name of topic for which to perform an election. Not allowed if --path-to-json-file or --all-topic-partitions is specified.")
+    .availableUnless("path-to-json-file")
+    .withRequiredArg
```
Should this be `requiredIf("partition")`?

jsancio replied:
That tried that at first. I haven't thought about the logic to implement something like this but unfortunately jopts-simple implementation of this required logic is limited. TL;DR; using `requiredIf` didn't work because I think there is a cycle in the graph.

hachikuji commented on core/src/main/scala/kafka/admin/LeaderElectionCommand.scala:
```diff
@@ -0,0 +1,235 @@
+      future.get()
+      false
+    } catch {
+      case (_: Throwable) =>
```
It may be useful to log this exception.

jsancio replied:
This is gone now.

jsancio reviewed (commented):
Thanks for the feedback. I'll address your comments today.

Regarding the server/controller side of the changes feel free to take a glance but hold off on a detail review. I found a major issue with the implementation that I have addressed and I am in the process of cleaning up.

cc @mumrah

hachikuji commented on core/src/main/scala/kafka/admin/LeaderElectionCommand.scala:
```diff
@@ -0,0 +1,235 @@
+      .map(topicPartition => topicPartition -> result.partitionResult(topicPartition))
+      .partition { case (_, partitionResult) => completedExceptionally(partitionResult) }
+
+    if (!succeeded.isEmpty) {
```
Unless I'm missing something, this doesn't work for the weird `partitions()` semantics, which returns successfully for all partitions which are specified explicitly regardless of the election result.

jsancio replied:
This all changed thanks to the much simpler API for `ElectLeadersResult`. Let me know if you still see logical errors.

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -635,38 +641,51 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
+    * Attempt to elect a replica as leader for each of the given partitions.
     * @param partitions The partitions to have their preferred leader elected
-    * @param electionType The election type
+    * @param ElectionType The type of election to perform
```
nit: should be `electionType`?

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -635,38 +641,51 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
         results.foreach { case (tp, throwable) =>
           if (throwable.isInstanceOf[ControllerMovedException]) {
-            error(s"Error completing preferred replica leader election for partition $tp because controller has moved to another broker.", throwable)
+            error(s"Error completing replica leader election ($electionType) for partition $tp because controller has moved to another broker.", throwable)
```
I wonder if this could be `info` since it does not indicate an actual problem.

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1549,14 +1568,22 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
+  case class ReplicaLeaderElection(
+    partitionsFromAdminClientOpt: Option[Set[TopicPartition]],
+    electionType: ElectionType,
+    electionTrigger: ElectionTrigger = ZkTriggered,
```
I wonder if we need to make this optional. Perhaps it's best for the caller to be explicit about the trigger?

jsancio replied:
Yeah. I am not a fan of default value in method arguments. I'll change this.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2389,32 +2401,31 @@ class KafkaApis(val requestChannel: RequestChannel,
+  def handleElectReplicaLeader(request: RequestChannel.Request): Unit = {
+
+    val electionRequest = request.body[ElectLeadersRequest]
+    val partitions: Set[TopicPartition] = if (electionRequest.data().topicPartitions() == null) {
```
I wonder if we ought to move this after the auth check so that we're less likely to inadvertently leak topic names.

jsancio replied:
Good catch. I also fixed the authorization branch so that we don't generate a map of partitions to error when the request version is after version 0.

hachikuji reviewed (commented):
Just a few more comments. The controller changes look pretty clean.

jsancio committed:
Propagate leader information during election

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/ElectLeadersRequest.java:
```diff
@@ -0,0 +1,132 @@
+
+        private ElectLeadersRequestData toRequestData(short version) {
+            if (electionType != ElectionType.PREFERRED && version == 0) {
+                throw new IllegalStateException("API Version 0 only supports PREFERRED election type");
```
I think we usually use `UnsupportedVersionException`.

junrao commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -635,38 +641,52 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
+    partitions: Set[TopicPartition],
+    electionType: ElectionType,
+    electionTrigger: ElectionTrigger
+  ): (Map[TopicPartition, Int], Map[TopicPartition, Throwable]) = {
```
Since we changed the return type, could we update the javadoc too?

jsancio replied:
We updated the return once more. I updated the comment to reflect the latest API.

junrao commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1600,23 +1628,30 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
+          debug(s"ReplicaLeaderElection waiting: $successfulPartitions, results: $results")
+          callback(
+            successfulPartitions,
+            results
```
Perhaps results can be renamed errorResults?

jsancio replied:
We changed the API once more. Now the result is not a tuple and it includes both the successful and failed elections.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1536,21 +1537,24 @@ class ReplicaManager(val config: KafkaConfig,
                          results: Map[TopicPartition, ApiError]): Unit = {
       if (expectedLeaders.nonEmpty) {
-        val watchKeys = expectedLeaders.map{
+        val watchKeys: Seq[TopicPartitionOperationKey] = expectedLeaders.map{
```
Perhaps results can be renamed errorResults?

jsancio replied:
We changed the API once more. Now the result is not a tuple and it includes both the successful and failed elections.

junrao commented on core/src/main/scala/kafka/controller/PartitionStateMachine.scala:
```diff
@@ -130,31 +135,36 @@ class ZkPartitionStateMachine(config: KafkaConfig,
+   * @param targetState The state
+   * @param partitionLeaderElectionStrategyOpt The leader election strategy if a leader election is required.
+   * @return a tuple with two maps when transition to OnlinePartition, for all other transtitions it returns two empty maps:
+   *         1. topic partitions and the corresponding expected leader ids
```
The comment probably needs to be adjusted a bit since uninitialized new partitions are not included.

junrao commented on core/src/main/scala/kafka/controller/PartitionStateMachine.scala:
```diff
@@ -391,24 +410,41 @@ class ZkPartitionStateMachine(config: KafkaConfig,
-    val (logConfigs, failed) = zkClient.getLogConfigs(partitionsWithNoLiveInSyncReplicas.map { case (partition, _) => partition.topic }, config.originals())
+
+    val (logConfigs, failed) = zkClient.getLogConfigs(
+      partitionsWithNoLiveInSyncReplicas.map { case (partition, _) => partition.topic },
```
If isUnclean is true, we can skip the reading of topic config from ZK.

jsancio replied:
Yes. Good catch. Fixed. I think we can simplify this method a bit more. I'll do this in a future PR.

junrao commented on core/src/main/scala/kafka/controller/PartitionStateMachine.scala:
```diff
@@ -391,24 +410,41 @@ class ZkPartitionStateMachine(config: KafkaConfig,
-    val (partitionsWithNoLiveInSyncReplicas, partitionsWithLiveInSyncReplicas) = leaderIsrAndControllerEpochs.partition { case (partition, leaderIsrAndControllerEpoch) =>
-      val liveInSyncReplicas = leaderIsrAndControllerEpoch.leaderAndIsr.isr.filter(replica => controllerContext.isReplicaOnline(replica, partition))
-      liveInSyncReplicas.isEmpty
+  private def collectUncleanLeaderElectionState(
```
This is not introduced in this PR, but it's useful to document the return type of the method.

jsancio replied:
Yeah. This method is difficult to follow. I added a comment. I'll try to make it easier to read in a future PR.

junrao commented on core/src/main/scala/kafka/controller/PartitionStateMachine.scala:
```diff
@@ -458,10 +494,10 @@ object PartitionLeaderElectionAlgorithms {
-case object ReassignPartitionLeaderElectionStrategy extends PartitionLeaderElectionStrategy
-case object PreferredReplicaPartitionLeaderElectionStrategy extends PartitionLeaderElectionStrategy
-case object ControlledShutdownPartitionLeaderElectionStrategy extends PartitionLeaderElectionStrategy
+final case class OfflinePartitionLeaderElectionStrategy(isUnclean: Boolean) extends PartitionLeaderElectionStrategy
```
It's probably more accurate to use allowUnclean instead of isUnclean?

junrao commented on core/src/main/scala/kafka/api/package.scala:
```diff
@@ -0,0 +1,58 @@
+import scala.collection.JavaConverters._
+import scala.collection.breakOut
+
+package object api {
```
It seems that it's a bit weird to name the scala file package.scala.

jsancio replied:
I think in Scala it is convention to name the file that contains package objects `package.scala`. E.g. `/path/to/package/package.scala`. Here are some examples:

1. Documentation: https://docs.scala-lang.org/tour/package-objects.html
2. Scala Library: https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/package.scala

junrao commented on core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala:
```diff
@@ -88,7 +92,50 @@ class PartitionStateMachineTest {
+     * Ending scenario: Leader: Y, Isr: [Y], Replicas: [X, Y], LiverBrokers: [Y]
+     *
+     * For the give staring scenario verify that performing an unclean leader
+     * election on the new partion results on the first live broker getting
```
typo partion

junrao commented on core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala:
```diff
@@ -251,7 +306,82 @@ class PartitionStateMachineTest {
+     * Ending scenario: Leader: Y, Isr: [Y], Replicas: [X, Y], LiverBrokers: [Y]
+     *
+     * For the give staring scenario verify that performing an unclean leader
+     * election on the offline partion results on the first live broker getting
```
typo partion

junrao commented on core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala:
```diff
@@ -88,7 +92,50 @@ class PartitionStateMachineTest {
+  }
+
+  @Test
+  def testNewPartitionToUncleanOnlinePartitionTransition(): Unit = {
```
Hmm, this test is weird since new partitions are elected through PartitionStateMachine.initializeLeaderAndIsrForPartitions and doesn't use OfflinePartitionLeaderElectionStrategy.

jsancio replied:
Yeah. I wasn't sure if I should add this. In theory, this should never happened in the controller implementation but the type don't enforce it. I'll remove it as to not confuse future readers.

junrao reviewed (commented):
@jsancio : Thanks for the PR. Made a pass of all the server side files. A few comments below.

jsancio committed:
Refactor controller to return leader election result

jsancio committed:
Add a top level error to the leader election RPC

jsancio committed:
Query all partitions after authorization

jsancio committed:
Only fetch topic config if allowUnclean is not set

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/ElectLeadersRequest.java:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    private final ElectLeadersRequestData data;
+    private final short version;
```
nit: we shouldn't need this since `AbstractRequest` already has a field for the version.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/ElectLeadersRequest.java:
```diff
@@ -0,0 +1,131 @@
+    public AbstractResponse getErrorResponse(int throttleTimeMs, Throwable e) {
+        ApiError apiError = ApiError.fromThrowable(e);
+        Map<String, Map<Integer, ApiError>> electionResult = new HashMap<>();
+        if (version == 0) {
```
In other cases, we set the error at both the top level and partition level. ListOffset is one example. Not sure if this is a pattern we need to follow here though.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/ElectLeadersRequest.java:
```diff
@@ -0,0 +1,131 @@
+     * Visible for testing.
+     */
+    @Override
+    public Struct toStruct() {
```
Looks like we can make this `protected` like the parent class and remove the "Visible for testing" comment.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/ElectLeadersResponse.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    public ElectLeadersResponse(Struct struct) {
+        this.version = (short) (ElectLeadersResponseData.SCHEMAS.length - 1);
```
nit: could use `ApiKeys.ELECT_LEADERS.latestVersion`

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/ElectLeadersResponse.java:
```diff
@@ -0,0 +1,155 @@
+
+    @Override
+    public boolean shouldClientThrottle(short version) {
+        return version >= 3;
```
Seems to be a bug from the original. I think we can just return `true` since client-side throttling was introduced before this API.

hachikuji commented on clients/src/main/resources/common/message/ElectLeadersResponse.json:
```diff
@@ -16,11 +16,13 @@
   "fields": [
     { "name": "ThrottleTimeMs", "type": "int32", "versions": "0+",
       "about": "The duration in milliseconds for which the request was throttled due to a quota violation, or zero if the request did not violate any quota." },
+    { "name": "ErrorCode", "type": "int16", "versions": "1+", "ignorable": false,
```
Do you think we may as well add an error message field as well?

jsancio replied:
I was wondering the same thing. I got inspiration from `FetchRequest`. There is no error message there so I figure I would not add it for consistency.

I'll add it if we think that every request should have `ThrottleTimeMs`, `ErrorCode` and `ErrorMessage` at the top level.

hachikuji replied:
I don't have a strong opinion. It would be nice to refactor the protocols to have a standard set of top level fields. Since we don't have a strong reason for the field now, maybe we can leave it for the future.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ElectLeadersResult.java:
```diff
@@ -0,0 +1,50 @@
+ * The API of this class is evolving, see {@link AdminClient} for details.
+ */
+@InterfaceStability.Evolving
+final public class ElectLeadersResult {
```
Might be convenient for users to have the `all()` API?

By the way, do you think it's worth fixing the issues with the old API? We can do this separately, just wondering if you think it's worthwhile.

jsancio replied:
Which issue specifically? That the user need to create `N` futures to get all of the errors?

If so, I think the user should just use `electLeaders` instead.

hachikuji replied:
Oh, I was referring to the buggy/inconsistent handling of the partitions=null case.

hachikuji replied:
Is it reasonable to add the `all()` API?

jsancio replied:
Sure. We have `partitions: Future[Map[TopicPartition, Optional[Throwable]]]`. Are you suggesting renaming this to `all`? Or do you want to add a new method `all: Future[Unit]` that throws if any of the partitions is set to a `Throwable`?

hachikuji replied:
I think it would be nice to have both. Many (maybe most?) of the AdminClient APIs offer `all()` for convenience.

hachikuji reviewed (commented):
Did another pass over the client-side changes. Will look at the server stuff next.

jsancio committed:
Implement electPreferredLeaders using electLeaders

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2943,34 +2944,83 @@ void handleFailure(Throwable throwable) {
+                        electionFuture.completeExceptionally(error.exception());
+                        return;
+                    }
+                } else if (topicPartitions == null && result.isEmpty()) {
```
I'm on the fence about the need to preserve this behavior in the new API. It seems weird to allow spurious error conditions. We'll be stuck with this for a long time. I wonder if we could just document the behavior. If a user requests to elect all partitions and no response is returned, then it could mean that the user is not authorized.

jsancio replied:
I don't have a strong opinion here. Are you thinking of just removing this `else if`?

Should the "kafka-preferred-leader-election" command handle this? Or just return the empty set there?

hachikuji replied:
Yeah, exactly. I just think the spurious auth error is more likely to cause confusion. This is even more crucial if we change the behavior of this API when partitions=null to only return the partitions that were successfully changed.

hachikuji commented:
In terms of error handling, I think these are the main cases:

1. Broker is not the controller => CONTROLLER_MOVED
2. Broker's zk write is fenced by another controller => Not 100% sure, but this raises a StateChangeFailedException which seems to get mapped to ELIGIBLE_LEADERS_NOT_AVAILABLE. Maybe this should be CONTROLLER_MOVED also?
3. Unclean/preferred election not needed => I was having trouble tracking this down, but I don't think we have a separate error code, so maybe it's ELIGIBLE_LEADERS_NOT_AVAILABLE? Should we have a new error code, maybe ELECTION_NOT_NEEDED?
4. Unclean/preferred election not possible => ELIGIBLE_LEADERS_NOT_AVAILBLE
5. Election succeeded => NONE

Just want to check high level if there are any additional cases to check for and if we agree from an API perspective on how they should be handled.

jsancio committed:
Do not assume cluster authorization error

Do not assume that if the admin client receives an empty set of
partitions when it did not provide any partitions that it means that we
had a cluster authoriazation error.

jsancio commented:
> In terms of error handling, I think these are the main cases:
> 
>     1. Broker is not the controller => CONTROLLER_MOVED
> 
>     2. Broker's zk write is fenced by another controller => Not 100% sure, but this raises a StateChangeFailedException which seems to get mapped to ELIGIBLE_LEADERS_NOT_AVAILABLE. Maybe this should be CONTROLLER_MOVED also?
> 
>     3. Unclean/preferred election not needed => I was having trouble tracking this down, but I don't think we have a separate error code, so maybe it's ELIGIBLE_LEADERS_NOT_AVAILABLE? Should we have a new error code, maybe ELECTION_NOT_NEEDED?
> 
>     4. Unclean/preferred election not possible => ELIGIBLE_LEADERS_NOT_AVAILBLE
> 
>     5. Election succeeded => NONE
> 
> 
> Just want to check high level if there are any additional cases to check for and if we agree from an API perspective on how they should be handled.

I filled the following issues. I will be working on them right after this PR.
1. https://issues.apache.org/jira/browse/KAFKA-8385
1. https://issues.apache.org/jira/browse/KAFKA-8384
1. https://issues.apache.org/jira/browse/KAFKA-8383

jsancio committed:
Change old command error semantic to mathc the new electLeaders RPC

jsancio committed:
Merge remote-tracking branch 'upstream/trunk' into election-rpc

jsancio committed:
Update documentation

jsancio committed:
Merge remote-tracking branch 'upstream/trunk' into election-rpc

jsancio committed:
Fix checkstyle

jsancio committed:
Either right bias is only supported in Scala 2.12

jsancio committed:
Add deprecation notice to the upgrade document

jsancio committed:
Change the admin test to now use deprecated APIs

hachikuji commented on core/src/main/scala/kafka/api/package.scala:
```diff
@@ -0,0 +1,58 @@
+      }
+    }
+
+    private[this] def byteToElectionType(value: Byte): ElectionType = {
```
nit: this seems like something which should be part of `ElectionType`

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1536,29 +1537,38 @@ class ReplicaManager(val config: KafkaConfig,
-                            partitions: Set[TopicPartition],
-                            responseCallback: Map[TopicPartition, ApiError] => Unit,
-                            requestTimeout: Long): Unit = {
+  def electLeaders(
```
There doesn't seem to be a good reason for this to be part of `ReplicaManager` instead of `KafkaController`. No need to do it here, but maybe we can move the election purgatory out of `ReplicaManager` in a follow-up patch.

jsancio replied:
https://issues.apache.org/jira/browse/KAFKA-8408

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -631,34 +637,49 @@ class KafkaController(val config: KafkaConfig,
   /**
-    * Attempt to elect the preferred replica as leader for each of the given partitions.
+    * Attempt to elect a replica as leader for each of the given partitions.
     * @param partitions The partitions to have their preferred leader elected
```
nit: comment still references preferred election

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1500,7 +1529,7 @@ class KafkaController(val config: KafkaConfig,
-          topicDeletionManager.isTopicQueuedUpForDeletion(partition.topic))
+            topicDeletionManager.isTopicQueuedUpForDeletion(partition.topic))
         if (partitionsBeingDeleted.nonEmpty) {
           warn(s"Skipping preferred replica election for partitions $partitionsBeingDeleted " +
```
A few messages like this still refer to preferred election

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1513,23 +1542,30 @@ class KafkaController(val config: KafkaConfig,
-          else
-            ApiError.fromThrowable(ex)
-          partition -> apiError
+        val results = onReplicaElection(electablePartitions, electionType, electionTrigger).mapValues {
```
Seems the criteria for `electablePartitions` only accounts for preferred leader election. For unclean leader election, this seems to work since the set of partitions eligible for unclean election is a subset, but maybe it is better to select the replicas explicitly?

jsancio replied:
Yeah. I am hesitant to make this change in this PR. As it would duplicate code in the PartitionStateMachine here and as a result creating technical debt. In the long term, I  want to refactor this code so that all of these decisions are made in the PartitionStateMachine which returns enough information for this layer to either filter those "errors" or to pass them along to the client.

What do you think?

hachikuji replied:
I was considering something simple like just collecting all the offline partitions.

hachikuji replied:
If we don't do this, then I think we won't return the `ELECTION_NOT_NEEDED` error consistently with the current logic. The only time you would get it would be for partitions whose preferred leader was already assigned, but that would be weird.

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1560,10 +1596,20 @@ class KafkaController(val config: KafkaConfig,
         case event: MockEvent =>
           event.process()
+        case ShutdownEventThread =>
+          /* This event is suppose to be handled by the ControllerEventThread. In the future we change the ControllerEventThread to handle:
```
Maybe we can file a JIRA and remove the comment?

jsancio replied:
https://issues.apache.org/jira/browse/KAFKA-8409

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1560,10 +1596,20 @@ class KafkaController(val config: KafkaConfig,
+           *
+           * and remove ShutdownEventThread from the ControlleveEvent enum.
+           */
+          error("Received a ShutdownEventThread event. This type of event is suppose to be handle by ControllerEventThread")
```
typo: supposed?

hachikuji reviewed (commented):
Thanks for the updates. Looking good overall. It would be nice to see some integration/system tests.

jsancio committed:
Don't return partitions that didn't need election

When the requesting elections for all of the topic partitions, do not
return topic partitions that didn't need to change leader.

jsancio committed:
Move the byte to ElectionType conversion function

jsancio committed:
Implement all() for ElectLeadersResult

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/ElectLeadersResponse.java:
```diff
@@ -0,0 +1,142 @@
+    public ElectLeadersResponse(
+            int throttleTimeMs,
+            short errorCode,
+            Map<String, Map<Integer, ApiError>> electionResults,
```
Would it make sense to specify this as `List<ReplicaElectionResult>`. It's a little annoying in `KafkaApis` to convert the scala type to a java type that we just have to convert again here.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2940,33 +2941,42 @@ void handleFailure(Throwable throwable) {
+                ElectLeadersResponse response = (ElectLeadersResponse) abstractResponse;
+                Map<TopicPartition, Optional<Throwable>> result = ElectLeadersResponse.electLeadersResult(response.data());
+
+                if (response.version() > 0) {
```
I think we could make `NONE` the default for this field and remove this check.

hachikuji commented on core/src/main/scala/kafka/admin/LeaderElectionCommand.scala:
```diff
@@ -0,0 +1,243 @@
+import scala.collection.mutable
+import scala.concurrent.duration._
+
+final object LeaderElectionCommand extends Logging {
```
nit: `final` is redundant with `object`

There are a few of these in this patch.

jsancio replied:
Removed.

Another habit. `object` are not always `final` it depends on where they are declared. See https://stackoverflow.com/questions/30265070/whats-the-point-of-nonfinal-singleton-objects-in-scala

hachikuji commented on core/src/main/scala/kafka/admin/LeaderElectionCommand.scala:
```diff
@@ -0,0 +1,243 @@
+    electionType: ElectionType,
+    topicPartitions: Option[Set[TopicPartition]]
+  ): Unit = {
+    val partitions = topicPartitions.map(_.asJava).getOrElse(null)
```
nit: `orNull`?

hachikuji commented on core/src/main/scala/kafka/admin/LeaderElectionCommand.scala:
```diff
@@ -0,0 +1,243 @@
+      ()
+    }
+
+    if (!succeeded.isEmpty) {
```
nit: use `nonEmpty`?

hachikuji commented on core/src/main/scala/kafka/admin/LeaderElectionCommand.scala:
```diff
@@ -0,0 +1,243 @@
+    } catch {
+      case e: ExecutionException =>
+        val cause = e.getCause
+        if (cause.isInstanceOf[TimeoutException]) {
```
nit: convert this to a `match`?

hachikuji commented on core/src/main/scala/kafka/admin/LeaderElectionCommand.scala:
```diff
@@ -0,0 +1,243 @@
+        succeeded += topicPartition
+      }
+
+      // Return unit. The scalac compiler throws an error because the expression above doesn't return Unit.
```
From the original, but the comment is not accurate and there has to be a nicer way to address this problem. Like maybe turn the block into an inner def with an explicit Unit type.

jsancio replied:
The other way is to do:
```
val _: Unit = if (...) {
  ...
} else {
  ...
}
```

hachikuji commented on core/src/main/scala/kafka/admin/LeaderElectionCommand.scala:
```diff
@@ -0,0 +1,243 @@
+    .describedAs("partition id")
+    .ofType(classOf[Integer])
+
+  val allTopicPartitions = parser
```
As far as I can tell, this is not used. We assume it is provided if the `singleTopicPartition` match above hits the  default case, but I'm wondering what the behavior is if the user specifies only --topic. Seems that would be allowed because it does not depend on --partition?

jsancio replied:
Yeah. It is use by jopt-simple to validate the flags but assumes it based on the constraints:
```
$ bin/kafka-leader-election.sh --topic test --bootstrap-server localhost:9092 --election-type unclean
Exception in thread "main" joptsimple.MissingRequiredOptionsException: Missing required option(s) [partition]
        at joptsimple.OptionParser.ensureRequiredOptions(OptionParser.java:426)
        at joptsimple.OptionParser.parse(OptionParser.java:400)
        at kafka.admin.LeaderElectionCommandOptions.<init>(LeaderElectionCommand.scala:237)
        at kafka.admin.LeaderElectionCommand$.run(LeaderElectionCommand.scala:46)
        at kafka.admin.LeaderElectionCommand$.main(LeaderElectionCommand.scala:42)
        at kafka.admin.LeaderElectionCommand.main(LeaderElectionCommand.scala)
```

I added this comment: https://github.com/apache/kafka/pull/6686/files#diff-d1bfa28bb2fe6969e0923a468b81e666R66

hachikuji commented on core/src/main/scala/kafka/admin/LeaderElectionCommand.scala:
```diff
@@ -0,0 +1,243 @@
+
+  val allTopicPartitions = parser
+    .accepts(
+      "all-topic-partitions",
```
The semantics we implement are really more like `all-eligible-partitions`. The name is fine, but might be worth emphasizing in the doc.

hachikuji commented on core/src/main/scala/kafka/admin/LeaderElectionCommand.scala:
```diff
@@ -0,0 +1,243 @@
+        throw e
+      case e: Throwable =>
+        println("Error while making request")
+        e.printStackTrace()
```
Not sure why this is not thrown like the other case. Maybe we can be consistent.

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1465,45 +1487,52 @@ class KafkaController(val config: KafkaConfig,
+  def electLeaders(
+    partitions: Set[TopicPartition],
+    electionType: ElectionType,
+    callback: ElectLeadersCallback =  _ => {}
```
Not sure if I asked about this already, but do we need the default callback?

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1465,45 +1487,52 @@ class KafkaController(val config: KafkaConfig,
       // leader elections and we get the `path exists` check for free
-      if (electionType == AdminClientTriggered || zkClient.registerZNodeChangeHandlerAndCheckExistence(preferredReplicaElectionHandler)) {
+      if (electionTrigger == AdminClientTriggered || zkClient.registerZNodeChangeHandlerAndCheckExistence(preferredReplicaElectionHandler)) {
         val partitions = partitionsFromAdminClientOpt match {
```
nit: could be `getOrElse(zkClient.getPreferredReplicaElection)`

By the way, I think we are assuming that `electionType` is PREFERRED if `partitionsFromAdminClientOpt` is not defined. This is true of course, but I'm wondering if we should validate the assumption.

jsancio replied:
Yeah. I prefer (no pun intended :smile:) to leave it as is. I think the core of the issue is that this method is trying to unify to many different behavior. In the future I want to split this method into at least two different methods. What do you think?

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1465,45 +1487,52 @@ class KafkaController(val config: KafkaConfig,
         val (validPartitions, invalidPartitions) = partitions.partition(tp => controllerContext.allPartitions.contains(tp))
         invalidPartitions.foreach { p =>
-          info(s"Skipping preferred replica leader election for partition ${p} since it doesn't exist.")
+          info(s"Skipping replica leader election ($electionType) for partition ${p} by $electionTrigger since it doesn't exist.")
```
nit: `${p}` -> `$p`. And maybe no harm spelling out "partition."

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1465,45 +1487,52 @@ class KafkaController(val config: KafkaConfig,
           case None => zkClient.getPreferredReplicaElection
         }
 
         val (validPartitions, invalidPartitions) = partitions.partition(tp => controllerContext.allPartitions.contains(tp))
```
The term `invalid` is ambiguous. Perhaps `nonExistingPartitions` would be clearer?

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1560,10 +1596,21 @@ class KafkaController(val config: KafkaConfig,
         case event: MockEvent =>
           event.process()
+        case ShutdownEventThread =>
+          /* See: KAFKA-8409
```
I prefer not cluttering the code with jira references. I think the suggestion below should just be part of the jira.

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1513,23 +1542,30 @@ class KafkaController(val config: KafkaConfig,
+          _ -> Left(new ApiError(Errors.UNKNOWN_TOPIC_OR_PARTITION, "The partition does not exist."))
+        )
+
+        debug(s"ReplicaLeaderElection waiting: $results")
```
Can we improve this message to clarify what exactly is being awaited?

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -631,34 +637,49 @@ class KafkaController(val config: KafkaConfig,
-          }
+      val strategy = electionType match {
+        case ElectionType.PREFERRED => PreferredReplicaPartitionLeaderElectionStrategy
+        case ElectionType.UNCLEAN => OfflinePartitionLeaderElectionStrategy(true)
```
Similar to a previous question, but we are assuming that we would hit this path only on admin client triggers. Perhaps it would be safer to either validate this or use:
```scala
OfflinePartitionLeaderElectionStrategy(allowUnclean = electionTrigger == AdminClientTriggered)
```
One thing I was considering as a possible future refactor is replacing the `UncleanLeaderElectionEnable` and `TopicUncleanLeaderElectionEnable` events with this general leader election event.

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -631,34 +637,49 @@ class KafkaController(val config: KafkaConfig,
+            } else {
+              error(s"Error completing replica leader election ($electionType) for partition $tp", throwable)
+            }
+          case (_, Right(_)) => // Ignored; No need to log or throw exception for the success cases
```
I guess we have a log message elsewhere?

jsancio replied:
I am reproducing the old code's behavior. Regarding logging, we would see a `LeaderAndIsrRequest` event from the `PartitionStateMachine`, right?

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1465,45 +1487,52 @@ class KafkaController(val config: KafkaConfig,
+    callback: ElectLeadersCallback
+  ): Unit = {
+    callback(partitionsFromAdminClientOpt.fold(Map.empty[TopicPartition, Either[ApiError, Int]]) { partitions =>
+        partitions.map(partition => partition -> Left(new ApiError(Errors.NOT_CONTROLLER, null)))(breakOut)
```
nit: seems misaligned

hachikuji commented on core/src/main/scala/kafka/controller/PartitionStateMachine.scala:
```diff
@@ -212,13 +227,18 @@ class ZkPartitionStateMachine(config: KafkaConfig,
-            stateChangeLog.trace(s"Changed partition $partition from ${partitionState(partition)} to $targetState with state " +
-              s"${controllerContext.partitionLeadershipInfo(partition).leaderAndIsr}")
-            controllerContext.putPartitionState(partition, OnlinePartition)
+          val electionResults = electLeaderForPartitions(partitionsToElectLeader, partitionLeaderElectionStrategyOpt.get)
```
I think the expectation is that an election strategy will be provided if the target state is online. Perhaps we could at least replace the call to `get` with a `getOrElse(throw ...)` with a clear message?

hachikuji commented on core/src/main/scala/kafka/controller/PartitionStateMachine.scala:
```diff
@@ -212,13 +227,18 @@ class ZkPartitionStateMachine(config: KafkaConfig,
+                s"Changed partition $partition from ${partitionState(partition)} to $targetState with state $leaderAndIsr"
+              )
+              controllerContext.putPartitionState(partition, OnlinePartition)
+            case (partition, Left(_)) => // Ignore; no need to update partition state on election error
```
nit: `case (_, Left(_))` since we are ignoring?

hachikuji commented on core/src/main/scala/kafka/controller/PartitionStateMachine.scala:
```diff
@@ -293,24 +313,30 @@ class ZkPartitionStateMachine(config: KafkaConfig,
+  ): Map[TopicPartition, Either[Throwable, LeaderAndIsr]] = {
     var remaining = partitions
-    var failures = Map.empty[TopicPartition, Throwable]
+    var finishedElections = Map.empty[TopicPartition, Either[Throwable, LeaderAndIsr]]
```
nit: may a well be an mutable map given the usage below

hachikuji commented on core/src/main/scala/kafka/controller/PartitionStateMachine.scala:
```diff
@@ -378,37 +412,81 @@ class ZkPartitionStateMachine(config: KafkaConfig,
+      }
+    }
+
+    (
```
nit: the parenthesis seems unneeded?

hachikuji commented on core/src/main/scala/kafka/controller/PartitionStateMachine.scala:
```diff
@@ -378,37 +412,81 @@ class ZkPartitionStateMachine(config: KafkaConfig,
+      }
+    } else {
+      val (logConfigs, failed) = zkClient.getLogConfigs(
+        partitionsWithNoLiveInSyncReplicas.map { case (partition, _) => partition.topic },
```
Should we convert this result to a set? Otherwise, I don't see what stops us from hitting zk multiple times for the same topic.

hachikuji commented on core/src/main/scala/kafka/controller/PartitionStateMachine.scala:
```diff
@@ -378,37 +412,81 @@ class ZkPartitionStateMachine(config: KafkaConfig,
-      controllerContext.partitionLeadershipInfo.put(partition, leaderIsrAndControllerEpoch)
-      controllerBrokerRequestBatch.addLeaderAndIsrRequestForBrokers(recipientsPerPartition(partition), partition,
-        leaderIsrAndControllerEpoch, replicas, isNew = false)
+    finishedUpdates.foreach { case (partition, result) =>
```
Would be really nice to move this logic to the caller, but not something we should try here.

hachikuji commented on core/src/main/scala/kafka/controller/ReplicaStateMachine.scala:
```diff
@@ -293,86 +299,116 @@ class ZkReplicaStateMachine(config: KafkaConfig,
+        } else None
+      }(breakOut)
+
+    val leaderIsrAndControllerEpochs = (leaderAndIsrsWithoutReplica ++ finishedPartitions).map { case (partition, result) =>
```
The types get a bit hard to follow in here...

hachikuji commented on core/src/main/scala/kafka/controller/ReplicaStateMachine.scala:
```diff
@@ -293,86 +299,116 @@ class ZkReplicaStateMachine(config: KafkaConfig,
+          case Some(leaderIsrAndControllerEpoch) =>
+            if (leaderIsrAndControllerEpoch.controllerEpoch > controllerContext.epoch) {
+              val exception = new StateChangeFailedException(
+                "Leader and isr path written by another controller. This probably" +
```
nit: missing space after "probably"

hachikuji commented on core/src/main/scala/kafka/controller/ReplicaStateMachine.scala:
```diff
@@ -293,86 +299,116 @@ class ZkReplicaStateMachine(config: KafkaConfig,
         }
       } else if (getDataResponse.resultCode == Code.NONODE) {
         partitionsWithNoLeaderAndIsrInZk += partition
+        None
```
The mix here of `flatMap` and mutable updates to `partitionsWithNoLeaderAndIsrInZk` make this a little hard to follow. Would it be simpler to make both collections mutable?

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2435,45 +2448,62 @@ class KafkaApis(val requestChannel: RequestChannel,
-          .setThrottleTimeMs(requestThrottleMs)
-          .setReplicaElectionResults(results.toList.asJava)
-        new ElectPreferredLeadersResponse(data)})
+        val adjustedResults = if (electionRequest.data().topicPartitions() == null) {
```
Might be worth a comment explaining this filtering.

hachikuji commented on core/src/main/scala/kafka/server/DelayedElectLeader.scala:
```diff
@@ -23,15 +23,16 @@ import org.apache.kafka.common.requests.ApiError
-                                  replicaManager: ReplicaManager,
-                                  responseCallback: Map[TopicPartition, ApiError] => Unit)
-    extends DelayedOperation(delayMs) {
+class DelayedElectLeader(
```
We don't have to do it here, but one thought I had is if it would be worthwhile to preempt the controller event if we timeout before completion.

jsancio replied:
Yeah. Let's do this in a future PR. I want to revisit how we manager network request/response and the controller event queue.

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1536,29 +1537,38 @@ class ReplicaManager(val config: KafkaConfig,
+    partitions: Set[TopicPartition],
+    electionType: ElectionType,
+    responseCallback: Map[TopicPartition, ApiError] => Unit,
+    requestTimeout: Long
```
The type here is `Long`, but the request actually specifies `Int`. Not sure it matters too much.

Another issue for a follow-up, but we take this value as is from the client without validation. This is probably a common pattern for admin APIs. Possibly we could enforce a range of values. At a minimum, it would be nice to take negative values out of the picture.

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1536,29 +1537,38 @@ class ReplicaManager(val config: KafkaConfig,
       if (expectedLeaders.nonEmpty) {
-        val watchKeys = expectedLeaders.map{
+        val watchKeys: Seq[TopicPartitionOperationKey] = expectedLeaders.map{
           case (tp, leader) => new TopicPartitionOperationKey(tp.topic, tp.partition)
```
nit: `new` not needed

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1536,29 +1537,38 @@ class ReplicaManager(val config: KafkaConfig,
-          new DelayedElectPreferredLeader(deadline - time.milliseconds(), expectedLeaders, results,
+        }(breakOut)
+        delayedElectLeaderPurgatory.tryCompleteElseWatch(
+          new DelayedElectLeader(deadline - time.milliseconds(), expectedLeaders, failures,
```
Not sure it matter, but might be a good idea to use `math.max(0, ...)` here

hachikuji reviewed (commented):
Thanks for the updates. Left some more comments.

jsancio committed:
Avoid unnecessary copy when generating response

jsancio committed:
Merge remote-tracking branch 'upstream/trunk' into election-rpc

jsancio committed:
For unclean filter partitions with a live leader

jsancio committed:
Use Set for collection of topics

jsancio committed:
Merge remote-tracking branch 'upstream/trunk' into election-rpc

jsancio committed:
It compiles in scala 2.11.x

jsancio commented:
retest this please

ijuma commented:
retest this please

jsancio committed:
Merge remote-tracking branch 'upstream/trunk' into election-rpc

jsancio commented:
retest this please

hachikuji reviewed (approved):
Thanks for the patch! LGTM.

jsancio committed:
Remove extra match case introduced during merge

jsancio committed:
Make a mutable copy of the topic Set

jsancio commented:
JDK 8 and Scala 2.11 passes. There is one failure in JDK 11 and Scala 2.12
> kafka.integration.UncleanLeaderElectionTest.testUncleanLeaderElectionDisabled

jsancio commented:
retest this please
